### ISSUE TITLE
Require scalaVersion

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#2795]
@tpolecat wrote https://twitter.com/tpolecat/status/789288986258931712:

> I think sbt should not start up unless you specify scalaVersion. Beginners think it uses the "installed" Scala which leads to suffering.

**Comment by eed3si9n** [comment]
This would break the existing behavior where the build would function without any build files, which allows things like sbt new. Additionally not all sbt projects are Scala projects.

**Comment by dwijnand** [comment]
Will have to workaround it for sbt new, but the plan is to require `sbt.version` in `project/build.properties` (possibly by prompting the user to confirm to create one).

With regards to non-Scala projects, I think perhaps we need to just leave scalaVersion undefined, which is fine for non-Scala projects.

**Comment by dwijnand** [comment]
Link to requiring sbt.version: #754 and #2668

**Comment by tpolecat** [comment]
Yeah, I realize there are some corner cases. But I think at least some kind of mechanism that checks for the version would help people out. I spent some time with a frustrated doobie user last night who honestly though he was using 2.11 but was actually using 2.10 (which requires the paradise plugin for some generic derivations but not others, so he was seeing random implicit not found errors). So 👍

**Comment by dwijnand** [comment]
Not defaulting scalaVersion would make it fail early. The question is how badly it explodes..

@tpolecat do you feel like trying it out for 1.0.x?

**Comment by soc** [comment]
I'd suggest starting without `scalaVersion`, but not providing Scala libraries by default. (Basically the opposite of `autoScalaLibrary := false`).
It would be a lot more consistent with other possible library dependencies like Guava, Apache Commons or Spring which are not added "just in case".

This would make it much easier for SBT to gain more adoption outside of the Scala ecosystem.

**Comment by dwijnand** [comment]
scala-library isn't added "just in case", it's added because sbt, even though sbt supports non-Scala projects, it started as a Scala build tool. So I think autoScalaLibrary defaulting to true is correct.

**Comment by soc** [comment]
Do we have no bigger ambitions than being a "Scala" build tool?
As an example, Gradle has been very successful not being a "Groovy" build tool.

I think some outside adoption is desirable, as it would encourage tooling providers to add better support for SBT and test codepaths which might not be taken that often in Scala.

This can only happen if non-Scala devs don't feel like second-class citizen right from the start.

**Comment by dwijnand** [comment]
It does, that's why I said started.

It's definitely used mostly for Scala projects over any other language, so defaulting to true makes sense.

**Comment by soc** [comment]
> It's definitely used mostly for Scala projects over any other language, so defaulting to true makes sense.

Isn't that confusing cause and effect?

**Comment by dwijnand** [comment]
I don't think sbt isn't used mostly for Java projects because of this default value.

Not having `autoScalaLibrary` at all would be detrimental to any uptake in other languages, but that's not the case here.

If over the next few years sbt successfully gains a huge number of non-Scala users then for 2.0 we should reconsider this choice. For now I don't think it's a good idea.

**Comment by soc** [comment]
> If over the next few years sbt successfully gains a huge number of non-Scala users then for 2.0 we should reconsider this choice.

I think it's sure as hell that this won't ever happen if non-Scala devs are treated as second-class citizens.

The market for "accept your second-class citizenship and maybe will make you feel more welcome 5 years down the road" seems to be quite small, given the existing competition.

**Comment by eed3si9n** [comment]
@soc As interesting as the idea of reducing Scala favoritism is, I think it's a whole different layer of discussion, so could you open a new issue or a thread on ML, please? On this issue, I want to explore the possibility of `scalaVersion`, which I think has much narrower scope.

**Comment by eed3si9n** [comment]
It might be helpful to display some information about the subproject when sbt starts up or when the subproject is switched could help. Something like:

```
[info] current subproject is zinc Core (in build file:/xxx/zinc/), Scala version 2.11.8
zinc Core>
```

**Comment by ekrich** [comment]
Not sure this is related but the default `scalaVersion` is `2.12.7` with sbt `1.2.6`. Should this be changed to `2.12.8`, the latest production version?

**Comment by dwijnand** [comment]
It's already changed for the next production version: https://github.com/sbt/sbt/pull/4509.  It just hasn't been released (unless you're running sbt 1.3.0-M1, but I don't recommend it seeing as both the supershell and `update` have some serious issues.)

**Comment by ekrich** [comment]
Thanks, @dwijnand

**Comment by eed3si9n** [comment]
If we think about this as removal of having an implicit `scalaVersion`, maybe I should start displaying warnings.

**Comment by SethTisue** [comment]
Currently (as of sbt 1.6.2) the default is 2.12.15. I would suggest:

* default to latest 2.13.x release at the time
  * (one could also argue for defaulting to latest 3.x)
* printing a warning when it does so

the only halfway good argument I can see for defaulting to 2.12.x is that that's the Scala version your build definition is written in. but I think that argument is.... only halfway good at best. it's 2022, 2.12.x is really old news. it's unfortunate (though understandable) that sbt build definitions are on 2.12, but I think beginners will very rarely notice whether their build definitions are on 2.12 or 2.13, and what matters more is the unfortunateness of steering them towards using a now quite old Scala version in their project code

**Comment by eed3si9n** [comment]
The sin of sbt 1.x, is that it was released in 2017, and sbt 2.x hasn't been developed. sbt 1.x should keep the build semantics, which means defaulting to 2.12.
If we ship sbt 2.x with Scala 3 in the build definition, the implicit scalaVersion will be Scala 3, which at some point will also be called old.
The only way to avoid being tied to an old Scala version, is to not allow blank Scala version as suggested by this issue.

**Comment by SethTisue** [comment]
I see. Well, a warning would be nice, at least.

**Comment by eed3si9n** [comment]
Here's a PR for this - https://github.com/sbt/sbt/pull/8428
